util/fmtutil/format.go

```diff
@@ -201,3 +199,102 @@ func makeLabelsMap(m *dto.Metric, metricName string, extraLabels map[string]stri
+
+	// Add histogram count timeseries.
+	labels[model.MetricNameLabel] = metricName + countStr
+	toTimeseries(wr, labels, timestamp, float64(h.GetSampleCount()))
+
+	// Add histogram sum timeseries.
+	labels[model.MetricNameLabel] = metricName + sumStr
+	toTimeseries(wr, labels, timestamp, h.GetSampleSum())
+}
+
+func addNativeHistogramToWriteRequest(wr *prompb.WriteRequest, h *dto.Histogram, labels map[string]string, timestamp int64) {
+	// Note that for float histograms:
+	// - Count should use CountFloat
+	// - NegativeCounts should bet used instead of NegativeDeltas
+	// - PositiveCounts should bet used instead of PositiveDeltas
+	//
+	// For float histograms:
+	// - Count should use CountInt
+	// - NegativeDeltas should bet used instead of NegativeCounts
+	// - PositiveDeltas should bet used instead of PositiveCounts
+	//
+	// *Counts and *Deltas are fine because the functions return nil if
+	// these are not set.
+	//
+	// The necessary interfaces for Count and ZeroCount are impossible to
+	// implement because they require a private method, this means it's not
+	// possible to create makeCount function that returns an instance of
+	// that interface.
+
+	hist := prompb.Histogram{
+		Sum:            h.GetSampleSum(),
+		Schema:         h.GetSchema(),
+		ZeroThreshold:  h.GetZeroThreshold(),
+		NegativeSpans:  makeBucketSpans(h.GetNegativeSpan()),
+		NegativeDeltas: h.GetNegativeDelta(),
+		NegativeCounts: h.GetNegativeCount(),
+		PositiveSpans:  makeBucketSpans(h.GetPositiveSpan()),
+		PositiveDeltas: h.GetPositiveDelta(),
+		PositiveCounts: h.GetPositiveCount(),
+		Timestamp:      timestamp,
+		// ResetHint: what is this?
+	}
+
+	if h.SampleCountFloat != nil {
+		// Assume float histogram
+		hist.Count = &prompb.Histogram_CountFloat{CountFloat: h.GetSampleCountFloat()}
+		hist.ZeroCount = &prompb.Histogram_ZeroCountFloat{ZeroCountFloat: h.GetZeroCountFloat()}
+	} else {
+		hist.Count = &prompb.Histogram_CountInt{CountInt: h.GetSampleCount()}
+		hist.ZeroCount = &prompb.Histogram_ZeroCountInt{ZeroCountInt: h.GetZeroCount()}
+	}
+
+	wr.Timeseries = append(wr.Timeseries, prompb.TimeSeries{
+		Labels:     makeLabels(labels),
+		Histograms: []prompb.Histogram{hist},
+	})
+}
+
+// isNativeHistogram returns false iff the provided histograms has no spans at
+// all (neither positive nor negative) and a zero threshold of 0 and a zero
+// count of 0. In principle, this could still be meant to be a native histogram
+// with a zero threshold of 0 and no observations yet. In that case,
+// instrumentation libraries should add a "no-op" span (e.g. length zero, offset
+// zero) to signal that the histogram is meant to be parsed as a native
+// histogram. Failing to do so will cause Prometheus to parse it as a classic
+// histogram as long as no observations have happened.
+func isNativeHistogram(h *dto.Histogram) bool {
+	return len(h.GetPositiveSpan()) > 0 ||
+		len(h.GetNegativeSpan()) > 0 ||
+		h.GetZeroThreshold() > 0 ||
+		h.GetZeroCount() > 0
+}
```

Comment:
I have been debating over how to handle this.

This function is a copy of the one found in [model/textparse/protobufparse.go](https://github.com/prometheus/prometheus/blob/83891135c6585fc2e9990824c9d12478acfe6b10/model/textparse/protobufparse.go#L575)

I would not mind the duplicate except that the chance of the original changing over time is non-zero.

One way is to play linker games. It's straightforward, but people might not like it.

The other is to move parts of `model/textparse` to an internal package (e.g. internal/model/textparse) and export this function from there so that this code can import it.

The other is to _export_ the function, sealing the contract regarding its future availability.

Any preference?

Reply:
tl;dr: I think we should stick with copying for now.

The problem is that the `fmtutil` version of this function (this one) imports `dto` as `dto "github.com/prometheus/client_model/go"`. The version in `textparse` umports `dto` as `dto "github.com/prometheus/prometheus/prompb/io/prometheus/client"`. The former is proto2 based and using the official Go protobuf library, the latter is proto3 based and using gogo protobuf. There is no difference on the wire, but the Go types are different.

The whole situation is very messy, but cleaning that up is not easy (and out of scope of this PR).